### ISSUE TITLE
Dynamic feature clustering

### ISSUE THREAD

**Issue by silvester-pari** [EOX-A/EOxElements#195]
Following [this OL example](https://openlayers.org/en/latest/examples/clusters-dynamic.html) and our implementation in [eodash](https://github.com/eurodatacube/eodash), it would be great to migrate this dynamic clustering as an option to the `eox-map`.
Ideally it should be able to cluster not only points, but any kind of features, if that makes sense; needs some brainstorming.

**Comment by santilland** [comment]
I think we need to divide the expected logic into more specific steps:

start with:

- [ ] grouping of point features when close together (based on some logic)

afterwards:

- [ ] click interaction 
   - [ ] zoom in to cluster extent
   - [ ] if points overlap, open as "spider" selection to allow selecting one of the points

I would propose to start with the grouping concept and afterwards see how we can manage interaction.
There is an inherited cluster layer from vector layer, so we will need a dedicated cluster layer type. It would be best to see if all interactions etc work on that.

**Comment by RobertOrthofer** [comment]
After some communication with @santilland , we detected a way to create the base-functionality of a clustered layer without the need of introducing new special Classes to the map element. Here is a basic `EOxLayer`:

```JSON
[
  {
    "type": "Vector",
    "properties": {
      "id": "clusterLayer"
    },
    "interactions": [
      {
        "type": "select",
        "options": {
          "id": "ClusterSelectInteraction",
          "condition": "click",
          "layer": {
            "type": "Vector",
            "properties": {
              "id": "clusterLayer"
            },
            "source": {
              "type": "Vector"
            },
            "style": {
              "stroke-color": "chartreuse",
              "stroke-width": 5
            }
          }
        }
      }
    ],
    "style": [{
        "filter": [">", ["get", "features", "length"], 1],
        "style": {
          "text-value": ["to-string", ["get", "features", "length"]],
          "circle-radius": 12,             
          "circle-fill-color": "#FFD700",  
          "circle-stroke-color": "rgba(255, 255, 255, 0.7)",
          "circle-stroke-width": 2,     
          "text-font": "normal 16px",         
          "text-fill-color": "#000",         
          "text-offset-y": 1
        }
      },
      {
        "else": true,
        "style": {
          "circle-radius": 12,             
          "circle-fill-color": "#FFD700",  
          "circle-stroke-color": "rgba(255, 255, 255, 0.7)",
          "circle-stroke-width": 2,     
          "text-font": "normal 16px",
          "text-value": "🗻",            
          "text-fill-color": "#000",         
          "text-offset-y": 1       
        }
      }],
    "source": {
      "type": "Cluster",
      "distance": 45,
      "source": {
        "type": "Vector",
        "url": "https://github.com/openlayers/openlayers/blob/main/examples/data/geojson/point-samples.geojson",
        "format": "GeoJSON"
      }
    }
  }
]
```

As this is a standard `VectorLayer`, The "zooming"-functionality might be doable utilizing the existing `Select` interaction, accessing the `features`-property of the clustered feature, and zooming the map if it's length of 2 or larger (= clustered feature).

As the "spider"-functionality seems not to be top priority right now, should we remove the `urgent`-tag from this issue?

**Comment by santilland** [comment]
Yes, we can remove the urgent tag, thank you for the example. We will try to integrate it like this, and we can come back to this issue more informed about what next steps we need.

**Comment by RobertOrthofer** [comment]
we can also implement the feature of clustering Polygons by using a `geometryFunction`, but this would have to be done on the side of the map element

**Comment by RobertOrthofer** [comment]
I agree that the the native cluster source is the correct tool for this behavior. Basically, what we need is two custom functionalities that we cannot replicate in our current EOxLayer-JSON:

 - geometry function for polygons
 - "spider-explode" functionality for clusters on click

I am not sure if we should allow drawing in the ClusterLayer directly, as a drawn point might be clustered immediately. I'm not sure if this is what the user would expect. We can still combine the [example above](https://github.com/EOX-A/EOxElements/issues/195#issuecomment-2703564892) with a "normal" VectorLayer without clustering, that switches to a clusterlayer after a certain zoom level, and only allow drawing above that zoom level.

For the custom functionalities, I'm thinking about simple keywords for the EoxLayer-definition, like `explodeOnClick: true` and `geometryFunction: 'extentCenter'`.
For consistency's sake, the `explodeOnClick` actually should be an `interaction`, just like a select-interaction. Not sure how this would interact with a normal click interaction, as the user would need to be able to click on the exploded points.

**Comment by silvester-pari** [comment]
I'd say if the user has zoomed out to a certain zoom level where the clustering threshold is reached, then it would be fine to immediately cluster when drawing. In this case the user would need to zoom in anyways to draw properly such a small feature.
In general, I'd say clustering should only happen for features that are too small to see, and thus are rendered as a point?

Regarding `explodeOnClick` I would argue that this would need to be the default anyways when using clustering, as otherwise there is no way to select things that are close to each other. So, like in our old eodash:
- if the points are on the exact same location, explode always
- if the points are close/overlapping and we're not yet at the highest zoom level, zoom in for the user to fit the points
- if the points are close/overlapping and we're already at the highest zoom level, explode

Regarding transformation of polygons to points, I'd say if the polygon becomes smaller than e.g. 10px*10px (so 100px on-screen area), turn it into a point and make it eligible for clustering. Any polygon above that will be shown as polygon and not be clustered. Ideally we can play around with the size threshold value to find a good balance.

Any opinions @santilland ?

**Comment by santilland** [comment]
Hi! I think your overview and the points you make on the behavior also previously in eodash are correct and make sense.
I can't think of anything to add. 

As Robert mentions i think we need to be careful about the interaction types:
> For consistency's sake, the explodeOnClick actually should be an interaction, just like a select-interaction. Not sure how this would interact with a normal click interaction, as the user would need to be able to click on the exploded points.

So, if we use the layer as source for feature selection in the draw tools, will this different interaction create conflict, etc
